pkg/api/types.go

```diff
@@ -228,21 +226,22 @@ type DcosConfig struct {
 
 // MasterProfile represents the definition of the master cluster
 type MasterProfile struct {
-	Count                    int         `json:"count"`
-	DNSPrefix                string      `json:"dnsPrefix"`
-	VMSize                   string      `json:"vmSize"`
-	OSDiskSizeGB             int         `json:"osDiskSizeGB,omitempty"`
-	VnetSubnetID             string      `json:"vnetSubnetID,omitempty"`
-	VnetCidr                 string      `json:"vnetCidr,omitempty"`
-	FirstConsecutiveStaticIP string      `json:"firstConsecutiveStaticIP,omitempty"`
-	Subnet                   string      `json:"subnet"`
-	IPAddressCount           int         `json:"ipAddressCount,omitempty"`
-	StorageProfile           string      `json:"storageProfile,omitempty"`
-	HTTPSourceAddressPrefix  string      `json:"HTTPSourceAddressPrefix,omitempty"`
-	OAuthEnabled             bool        `json:"oauthEnabled"`
-	PreprovisionExtension    *Extension  `json:"preProvisionExtension"`
-	Extensions               []Extension `json:"extensions"`
-	Distro                   Distro      `json:"distro,omitempty"`
+	Count                    int              `json:"count"`
+	DNSPrefix                string           `json:"dnsPrefix"`
+	VMSize                   string           `json:"vmSize"`
+	OSDiskSizeGB             int              `json:"osDiskSizeGB,omitempty"`
+	VnetSubnetID             string           `json:"vnetSubnetID,omitempty"`
+	VnetCidr                 string           `json:"vnetCidr,omitempty"`
+	FirstConsecutiveStaticIP string           `json:"firstConsecutiveStaticIP,omitempty"`
+	Subnet                   string           `json:"subnet"`
+	IPAddressCount           int              `json:"ipAddressCount,omitempty"`
+	StorageProfile           string           `json:"storageProfile,omitempty"`
+	HTTPSourceAddressPrefix  string           `json:"HTTPSourceAddressPrefix,omitempty"`
+	OAuthEnabled             bool             `json:"oauthEnabled"`
+	PreprovisionExtension    *Extension       `json:"preProvisionExtension"`
+	Extensions               []Extension      `json:"extensions"`
+	Distro                   Distro           `json:"distro,omitempty"`
+	KubernetesConfig         KubernetesConfig `json:"kubernetesConfig,omitempty"`
```

Comment:
Is KubernetesConfig part of both OrchestratorProfile and Master+Agent Profile?

Reply:
For the purposes of this PR, this is a change that enables master-specific and agent-specific kubelet configs (as opposed to using a common one).

In a longer-term sense, adding `KubernetesConfig` to the master/agents separately means that we can do k8s configuration generally on a per-master and per-agent node basis.

I couldn't find a more elegant way to do this from a kubelet config perspective, but I'm open to ideas.

Reply:
Ok that makes sense. What about removing `KubernetesConfig` from `OrchestratorProfile`? In that case `KubernetesConfig` would act like `vmSize` where masters and agents have their own and in the case of a common one it is duplicated in the master and agent profile.

Reply:
Eventually we may do that, but that's pretty destructive (everyone's api models out there in the world refer to it). My thoughts for this are:

- `OrchestratorProfile.KubernetesConfig` = common (all Azure resourses) Kubernetes-related config
- `MasterProfile.KubernetesConfig` = Kubernetes-related config for master nodes
- `AgentPoolProfile.KubernetesConfig` = Kubernetes-related config for agent nodes in an agent pool

Reply:
So what happens if I use the following api model? 

`{
  "apiVersion": "vlabs",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes",
      "kubernetesConfig": {
          "clusterSubnet": "10.230.0.0/16"
      }
    },
    "masterProfile": {
      "count": 1,
      "dnsPrefix": "cerobert1218-etcd3",
      "vmSize": "Standard_D2_v2",
    },
    "agentPoolProfiles": [
      {
        "name": "agentpool1",
        "count": 3,
        "vmSize": "Standard_D2_v2",
        "availabilityProfile": "AvailabilitySet"
      }
    ],`

Reply:
So what happens if I use the following api model? 

{
  "apiVersion": "vlabs",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes",
      "kubernetesConfig": {
        "clusterSubnet": "10.230.0.0/16"
      }
    },
    "masterProfile": {
      "count": 1,
      "dnsPrefix": "",
      "vmSize": "Standard_D2_v2",
"kubernetesConfig": {
        "clusterSubnet": "10.230.0.0/16"
      }
    },
    "agentPoolProfiles": [
      {
        "name": "agentpool1",
        "count": 3,
        "vmSize": "Standard_D2_v2",
        "availabilityProfile": "AvailabilitySet",
      }
    ],
...
}

Reply:
At present, nothing. The operable changes are here:

https://github.com/Azure/acs-engine/pull/1854/files#diff-8edb2653431a5da9ad1f4286097702a7R479

That copies over the currently supported KubernetesConfig (from OrchestratorProfile) so that it can be consumed thusly:

by the agents:
- https://github.com/Azure/acs-engine/pull/1854/files#diff-7d5f50f227edfc447ff04539035aa614R106
by the masters:
- https://github.com/Azure/acs-engine/pull/1854/files#diff-91eb315fd90ae6b7177cd6290a360600R235